qpid/cpp/src/qpid/broker/amqp/Outgoing.cpp

```diff
@@ -142,11 +145,10 @@ void OutgoingFromQueue::handle(pn_delivery_t* delivery)
                 if (preAcquires()) {
                     //TODO: handle message-annotations
                     if (pn_disposition_is_undeliverable(pn_delivery_remote(delivery))) {
-                        //treat undeliverable here as rejection
-                        queue->reject(r.cursor);
-                    } else {
-                        queue->release(r.cursor, pn_disposition_is_failed(pn_delivery_remote(delivery)));
+                        undeliverableMessages.add(r.msg.getSequence());
                     }
+
+                    queue->release(r.cursor, pn_disposition_is_failed(pn_delivery_remote(delivery)));
```

Comment:
One issue here is that the message will remain on the queue, even if there are no other consumers. In some cases that may be what you want, in others though it could cause a problem.

What would you think of having a queue level setting that dictated whether consumers should reject or track undelivered? Ideally we would then have reject as the default for now for backwards compatibility, but the behaviour you desire could be enabled when creating the queue? Alternatively we could even make the default a broker level option.